occ web executor

VicDeo opened the pull request:
As per https://github.com/owncloud/core/issues/21985

VicDeo committed:
Initial web executor

DeepDiver1975 commented:
Nice! Go for it!

mmattel commented:
**Reminder:**
We must not forget to check Apache/nginx rules to allow web access for the occ command (according to #21985 ``https://domain.tld/owncloud/index.php/occ/app:list?format=json``).

Maybe also a check in core if web access to occ is possible and log this - else we never find why it is not working...

Possible changes needed:
- Apache (in core)
```
 RewriteRule ^(\.|autotest|occ|issue|indie|db_|console).* - [R=404,L]
```
- nginx (in the documentation):
```
location ~ ^/(?:\.|autotest|occ|issue|indie|db_|console) {
        deny all;
    }
```

nickvergessen commented:
Woot index.php/occ is called, not the occ file directly?

PVince81 commented:
> Woot index.php/occ is called, not the occ file directly?

Probably because the ajax request needs to be translated to actual arguments.
Might work as well to call the "occ" one directly but then you lose the advantage of routing / parameter mapping.

nickvergessen commented:
> Probably because the ajax request needs to be translated to actual arguments.
Might work as well to call the "occ" one directly but then you lose the advantage of routing / parameter mapping.

Well you shouldn't call it directly, because it's a bash file ;) But that is exactly what I wanted to point out. The occ file is still not needed to be accessible, so why should any rewrite rule require changing?

mmattel commented:
@nickvergessen you are right.
I have overseen that ``https://domain.tld/owncloud/index.php/occ/app:list?format=json`` does not access occ via ownclouds root as a shell script. Root access is covered and secured by the rewrite rules. Sorry for the confusion.  :smile:

DeepDiver1975 commented:
@VicDeo what's the status with this? THX

PVince81 commented:
Code already looks good so far

nickvergessen commented on core/Controller/OccController.php:
```diff
@@ -0,0 +1,130 @@
+		'app:getpath',
+		'app:list',
+		'check',
+		'config:list',
```
can we check on the parameters? `--private` should not be allowed, because otherwise you can get the plaintext database, mail and other passwords.

VicDeo replied:
@nickvergessen `--private` is required to read `updater.secret` see https://github.com/owncloud/updater/pull/233

VicDeo commented:
@PVince81 The last step is 
- [ ] Bypass maintenance mode

PVince81 commented:
@VicDeo when you change labels please post a comment. Changing labels doesn't trigger a github notification.

So this is ready for review ?

nickvergessen commented on core/Controller/OccController.php:
```diff
@@ -0,0 +1,131 @@
+		'upgrade'
+	];
+
+	public function __construct($appName, IRequest $request) {
```
just leave this out?

nickvergessen commented on core/Controller/OccController.php:
```diff
@@ -0,0 +1,131 @@
+			);
+		}
+
+		if (!password_verify($token, $coreToken)) {
```
Not available on 5.4 http://php.net/manual/en/function.password-verify.php

VicDeo replied:
@nickvergessen that's why we have https://github.com/owncloud/updater/blob/master/vendor/ircmaxell/password-compat/lib/password.php in

VicDeo replied:
oops, wrong link. I mean  3rdparty https://github.com/owncloud/3rdparty/blob/master/ircmaxell/password-compat/lib/password.php

nickvergessen commented on lib/base.php:
```diff
@@ -271,9 +271,13 @@ public static function checkInstalled() {
+		if (
+			\OC::$server->getSystemConfig()->getValue('maintenance', false)
+			&& OC::$SUBURI != '/core/ajax/update.php'
+			&& strpos($request->server['PATH_INFO'], '/occ/') !== 0
```
this is a bit hardcore, it will ignore maintenance mode to any folder, etc that is named occ

nickvergessen replied:
Use the following instead:
```php
!(substr($request->getScriptName(), 0 - strlen('/index.php')) === '/index.php'
&& strpos($request->getPathInfo(), '/occ/') === 0)
```

nickvergessen commented on lib/base.php:
```diff
@@ -271,9 +271,13 @@ public static function checkInstalled() {
 	}
 
-	public static function checkMaintenanceMode() {
+	public static function checkMaintenanceMode($request) {
```
missing type hint

VicDeo commented:
@nickvergessen thanks, everything's addressed

VicDeo commented:
@PVince81 yes, it's ready for review. [Related updater PR](https://github.com/owncloud/updater/pull/355) is still under development though

PVince81 commented on core/Controller/OccController.php:
```diff
@@ -0,0 +1,127 @@
+			$inputArray = array_merge(['command' => $command], $params);
+			$input = new ArrayInput($inputArray);
+
+			$exitCode = $application->run($input, $output);
```
I wonder whether we should blacklist/whitelist commands. But the fact that the caller needs the secret + localhost is probably already enough

nickvergessen replied:
there is already a whitelist in the beginning of `core/Controller/OccController.php`?

VicDeo replied:
@nickvergessen @PVince81 yes, there is a whitelist https://github.com/owncloud/core/pull/24957/files#diff-1b29c3a8398e3ed87e772fb50a729633R110
Only commands required for updater are allowed for now.

PVince81 commented:
The code looks good and makes sense :+1: 

Interestingly this means that regular apps could also use the OccController to run stuff, but they'd then also need to grab the secret from the config.

nickvergessen commented:
> Interestingly this means that regular apps could also use the OccController to run stuff, but they'd then also need to grab the secret from the config.

Well they could also implement it themselves and then even without the secret :speak_no_evil:

PVince81 commented:
@DeepDiver1975 can you review ?

I also realized that this is something we'll likely want to backport to 9.0 at some point (including the updater part) to make sure people who use the updater in 9.0 without proc_open support can still update properly.

DeepDiver1975 commented:
OccController has no unit tests - this has to be added please

VicDeo commented:
@DeepDiver1975 done

DeepDiver1975 commented:
@VicDeo integration tests look fishy - please have a look

````
21:23:59 [Mon Jun 20 21:24:55 2016] Argument 1 passed to OC::checkMaintenanceMode() must implement interface OCP\IRequest, none given, called in /ssd/jenkins/workspace/ocs-api-integration-tests-ci/public.php on line 38 and defined at /ssd/jenkins/workspace/ocs-api-integration-tests-ci/lib/base.php#280
21:23:59 [Mon Jun 20 21:24:55 2016] Undefined variable: request at /ssd/jenkins/workspace/ocs-api-integration-tests-ci/lib/base.php#282
21:23:59 [Mon Jun 20 21:24:55 2016] PHP Fatal error:  Call to a member function getScriptName() on null in /ssd/jenkins/workspace/ocs-api-integration-tests-ci/lib/base.php on line 282
21:23:59 [Mon Jun 20 21:24:55 2016] PHP Stack trace:
21:23:59 [Mon Jun 20 21:24:55 2016] PHP   1. {main}() /ssd/jenkins/workspace/ocs-api-integration-tests-ci/public.php:0
21:23:59 [Mon Jun 20 21:24:55 2016] PHP   2. OC::checkMaintenanceMode() /ssd/jenkins/workspace/ocs-api-integration-tests-ci/public.php:38
21:23:59 [Mon Jun 20 21:24:55 2016] ::1:56378 [500]: /public.php/webdav/ - Call to a member function getScriptName() on null in /ssd/jenkins/workspace/ocs-api-integration-tests-ci/lib/base.php on line 282
21:23:59 [Mon Jun 20 21:24:55 2016] Call to a member function getScriptName() on null at /ssd/jenkins/workspace/ocs-api-integration-tests-ci/lib/base.php#282
21:23:59 [Mon Jun 20 21:24:55 2016] Internal Server Error

DeepDiver1975 committed:
Fix PHPDoc

Fix broken integration test

OccControllerTests do not require database access - moch them all!

Kill unused sprintf

DeepDiver1975 commented on tests/Core/Controller/OccControllerTest.php:
```diff
@@ -0,0 +1,131 @@
+	private $controller;
+
+	public static function setUpBeforeClass(){
+		self::$oldSecret = \OC::$server->getConfig()->setSystemValue('updater.secret', '');
```
@VicDeo  please inject an instance of IConfig into the OccController and mock this onject for testing. THX

DeepDiver1975 commented on lib/base.php:
```diff
@@ -271,9 +273,20 @@ public static function checkInstalled() {
+	 * Limit maintenance mode access
+	 * @param IRequest $request
+	 */
+	public static function checkMaintenanceMode(IRequest $request) {
```
this is breaking the integration test :boom:

DeepDiver1975 commented on core/Controller/OccController.php:
```diff
@@ -0,0 +1,127 @@
+	 */
+	protected function validateRequest($command, $token){
+		if (!in_array($this->request->getRemoteAddress(), ['::1', '127.0.0.1', 'localhost'])) {
+			throw new \UnexpectedValueException(sprintf('Web executor is not allowed to run from a different host', $command));
```
unneccessary sprintf

DeepDiver1975 commented on tests/Core/Controller/OccControllerTest.php:
```diff
@@ -0,0 +1,133 @@
+	}
+
+	private function getRequestMock($host){
+		$this->request = $this->getMockBuilder('OC\AppFramework\Http\Request')
```
no need to mock the occcontroller

DeepDiver1975 commented on tests/Core/Controller/OccControllerTest.php:
```diff
@@ -0,0 +1,133 @@
+
+	private function getControllerMock($host){
+		$this->getRequestMock($host);
+		$this->controller = $this->getMockBuilder('OC\Core\Controller\OccController')
```
no need to mock the occcontroller

VicDeo commented on tests/Core/Controller/OccControllerTest.php:
```diff
@@ -88,6 +86,14 @@ public function testWrongToken(){
+			->willReturnCallback(
+				function ($input, $output) {
+					/** @var Output $output */
+					$output->writeln('{"installed":true,"version":"9.1.0.8","versionstring":"9.1.0 beta 2","edition":""}');
```
@DeepDiver1975 I didn't go this way because it requires adjustment with any version bump.

DeepDiver1975 replied:
Not really relevant because we don't assert on the version.

DeepDiver1975 commented:
:+1:

PVince81 commented:
stable9: https://github.com/owncloud/core/issues/25222
